File: 프로그래머스/2/131127. 할인 행사/할인 행사.py
def solution(want, number, discount):
    answer = 0
    want_info = dict()
    N = len(want)
    for i in range(N):
        want_info[want[i]] = number[i]
    
    M = len(discount)
    
    for i in range(M-9):
        tmp = dict()
        possible = True
        for j in range(10):
            if discount[i+j] not in want:
                possible = False
                break
                
            else:
                if discount[i+j] in tmp:
                    tmp[discount[i+j]] += 1
                    if tmp[discount[i+j]] > want_info[discount[i+j]]:
                        possible = False
                        break
                else:
                    tmp[discount[i+j]] = 1
        if possible:
            for k, v in want_info.items():
                if tmp[k] != v:
                    possible = False
                    break
                    
        if possible:
            answer += 1
        
    return answer

File: 프로그래머스/2/131127. 할인 행사/할인 행사.java
import java.util.*;
class Solution {
    public int solution(String[] want, int[] number, String[] discount) {
        int answer = 0;
        Map<String, Integer> map = new HashMap<>();
        int N = want.length;
        int M = discount.length;
        for (int i = 0; i < N; i++) {
            map.put(want[i], number[i]);
        }
        
        for (int i = 0; i < M-9; i++) {
            Map<String, Integer> tmp = new HashMap<>();
            boolean possible = true;
            for (int j = 0; j < 10; j++) {
                if (!map.containsKey(discount[i+j])) {
                    possible = false;
                    break;
                }
                
                tmp.put(discount[i+j], tmp.getOrDefault(discount[i+j], 0) + 1);
                if (tmp.get(discount[i+j]) > map.get(discount[i+j])) {
                    possible = false;
                    break;
                }
            }
            if (possible) {
                for (String key : map.keySet()) {
                    if (map.get(key) != tmp.get(key)) {
                        possible = false;
                        break;
                    }
                }
            }
            if (possible) answer++;
        }
        return answer;
    }
}